File: backend/app/schemas/scans.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

ScanStatus = Literal["pending", "running", "completed", "failed", "cancelled"]


class ScanCreate(BaseModel):
    # None = scan all enabled roots.
    root_ids: list[int] | None = None


class ScanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: ScanStatus
    root_ids: list[int] | None
    files_found: int
    files_processed: int
    files_added: int
    files_changed: int
    files_unchanged: int
    files_missing: int
    files_moved: int
    error_count: int
    current_path: str | None
    message: str | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime


class ScanErrorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    path: str
    error: str
    created_at: datetime


class ScanErrorPage(BaseModel):
    items: list[ScanErrorRead]
    total: int
    limit: int
    offset: int

File: frontend/src/api/scans.ts
import { request, requestJson } from './client'

export type ScanStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

export interface ScanRead {
  id: number
  status: ScanStatus
  root_ids: number[] | null
  files_found: number
  files_processed: number
  files_added: number
  files_changed: number
  files_unchanged: number
  files_missing: number
  files_moved: number
  error_count: number
  current_path: string | null
  message: string | null
  started_at: string | null
  finished_at: string | null
  created_at: string
}

export interface ScanError {
  id: number
  path: string
  error: string
  created_at: string
}

export interface ScanErrorPage {
  items: ScanError[]
  total: number
  limit: number
  offset: number
}

export const TERMINAL_SCAN_STATUSES: ScanStatus[] = ['completed', 'failed', 'cancelled']

export function startScan(rootIds: number[] | null): Promise<ScanRead> {
  return requestJson<ScanRead>('/api/scans', 'POST', { root_ids: rootIds })
}

export function getScan(id: number): Promise<ScanRead> {
  return request<ScanRead>(`/api/scans/${id}`)
}

export function listScans(limit = 1): Promise<ScanRead[]> {
  return request<ScanRead[]>(`/api/scans?limit=${limit}`)
}

export function cancelScan(id: number): Promise<ScanRead> {
  return requestJson<ScanRead>(`/api/scans/${id}/cancel`, 'POST', {})
}

/** Files the scan could not read — unreadable, corrupt, or permission-denied. */
export function listScanErrors(scanId: number, limit = 100): Promise<ScanErrorPage> {
  return request<ScanErrorPage>(`/api/scans/${scanId}/errors?limit=${limit}`)
}
